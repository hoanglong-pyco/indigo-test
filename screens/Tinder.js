import React, { useState, useEffect } from "react";
import { ActivityIndicator, View } from "react-native";
import * as favorite from "../sqlite/Favorite";

import styled from "styled-components/native";

import * as personApi from '../api/person';
import PersonSwipe from "../components/PersonSwipe";

// Schedule preload items
class Schedule extends Array {
  constructor(props) {
    super();
    this.props = Object.assign(
      {
        preload: 1,
        load: () => Promise.resolve(),
        onLoad: Boolean
      },
      props
    );
    this.preload();
  }
  shift() {
    const item = super.shift();
    this.preload();
    return item;
  }

  //calculate and preload items if needed.
  async preload() {
    const { preload, load, onLoad } = this.props;
    if (this.length > preload) return this;
    const item = await load();
    this.push(item);
    onLoad.call(this, item);
    return this.preload();
  }
}

export default function Tinder({}) {
  const [schedule, setSchedule] = useState(null);
  const [persons, setPersons] = useState([]);

  useEffect(() => {
    const schedule = new Schedule({
      preload: 2, // api will preload 2 person
      onLoad: () => persons.length || setPersons(schedule.slice(0,1)), // Only set one first item on schedule
      load: personApi.random
    });
    setSchedule(schedule);
  }, []);

  return (
    <Wrapper>
      <View>
        {persons[0] ? (
          persons.map((person, index) => (
            <View
              key={person.md5}
              style={{ position: "absolute", left: -140, top: -150 }}
            >
              <PersonSwipe
                person={person}
                onSwipeLeft={person => console.log("Next", person)}
                onSwipeRight={favorite.insert}
                onDrop={() => {
                  // remove first items and reload new item
                  schedule.shift();
                  setPersons(schedule.slice(0,1));
                }}
              />
            </View>
          ))
        ) : (
          <ActivityIndicator />
        )}
      </View>
    </Wrapper>
  );
}

Tinder.navigationOptions = {
  title: "Tinder"
};

const Wrapper = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
`;
