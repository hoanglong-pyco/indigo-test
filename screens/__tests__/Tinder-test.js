import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";

import Tinder from "../Tinder";

const responseData = {
  results: [
    {
      user: {
        gender: "female",
        name: { title: "ms", first: "marilyn", last: "welch" },
        location: {
          street: "7510 e little york rd",
          city: "brigham city",
          state: "oregon",
          zip: "46975"
        },
        email: "marilyn.welch75@example.com",
        username: "greendog816",
        password: "lookin",
        salt: "hXycKiXy",
        md5: "7ca60bdada260ca958a33230c43bf358",
        sha1: "a0fe616659f16fcae9c663aa7aa56e3e833824da",
        sha256:
          "6b8e15325810d927461ada6034a5136451515c2e3f50309d5e7512319f405ede",
        registered: "948870895",
        dob: "406521725",
        phone: "(143)-190-7590",
        cell: "(291)-287-7518",
        SSN: "405-54-5739",
        picture: "http://api.randomuser.me/portraits/women/80.jpg"
      },
      seed: "a743dcfc7d6311f1",
      version: "0.4"
    }
  ]
};

it(`renders Tinder`, () => {
  global.fetch = jest.fn(() => () => Promise.resolve({ json: () => Promise.resolve(responseData) }));
  const person = renderer.create(<Tinder />).toJSON();
  expect(person).toMatchSnapshot();
});
