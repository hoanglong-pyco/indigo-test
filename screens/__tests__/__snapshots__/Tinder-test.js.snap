// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Tinder 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View>
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
</View>
`;
